text/0069-api-v2.md

```diff
@@ -0,0 +1,172 @@
+
+1. RawKV is not safe to use along with TxnKV. By solving this, TiDB will be able to support RawKV as the table's storage engine, which will enrich TiDB's use cases.
+2. RawKV TTL is controlled by Store configuration. Switching the configuration will cause data corruption in silence.
+3. RawKV TTL is encoded into the value by appending 8-bytes UNIX timestamp to the end of the value, therefore it's hard to introduce other encode afterward.
+4. RawKV compare_and_swap is easy to make mistake.
+5. It could be nice if we can deploy multiple applications on one TiKV cluster.
+
+## Detailed Design
+
+#### New key-value codec
+
+This RFC introduces a new key encode to RawKV and TxnKV, and a new value encode to RawKV, which will allow the RawKV to be used along with TxnKV and also allow TiKV to flexibly add meta, e.g. the TTL, to a RawKV value.
+
+In addition, keys will be contained in keyspaces, where the keys in different keyspace are totally independant. If keyspace is not specified, the keyspace 'default' will be used.
+
+The `API V2` is enabled by a switch on PD. Since it changes the storage encode, it will be not compatible to switch between `API V1` and `API V2` while there are non-TiDB data in TiKV. TiDB data is specially treated in order to not be affected by the change.
+
+#### Key Encode
+
+Once the `API V2` is enabled, the key will be starting with either:
+
+1. `m` or `t`: TxnKV key. Used by TiDB.
+2. `x{keyspace prefix id}`: TxnKV key.
+3. `r{keyspace prefix id}`: RawKV key.
+4. `a{keyspace prefix id}`: RawKV atomic key, on which you can use `CompareAndSwap`.
+
+The `{keyspace prefix id}` is the [keyspace](https://github.com/tikv/rfcs/pull/39) prefix for seperating keys of different keyspace. It should be an vary-length integer whose highest bit of every byte denotes whether the next byte is still part of the integer. The client will fetch the prefix from PD by the keyspace name specified by the user when initializing the client, so it means that the keyspace prefix is valid during the session, in other words, the change on keyspace on PD will not take effect on running seesions.
+
+In addition, the user key in every mode will be encoded by `Memory Comparable Encoding` as TxnKV does in API V1.
+
+#### RawKV Value Encode
+
+If the key is started with `r` or `a`, then the value can be either:
+
+1. `{0x0}{data}`
+2. `{0x1}{TTL expire timestamp}{data}`
+
+### Keyspace Management
+
+Add a new http interface to PD for adding, renaming, deleteing and querying the mapping from keyspace name to prefix id:
+
+```javascript
+// list all keyspaces
+GET /keyspaces
+[
+  {
+    name: "default",
+    id: "0"
+  },
+  {
+    name: "redis",
+    id: "1"
+  }
+]
+
+// add new keyspace
+POST /keyspaces
+{
+  name: "mongodb",
+  // optional, pd will allocate one if not specified
+  id: "2"
+}
+
+// delete a keyspace
+DELETE /keyspaces/{keyspace_name}
```

Comment:
No, the client no longer have access to the deleted keyspace.

Reply:
when the client is firstly connected to the cluster, it will record the mapping between keyspace name and id, how to notify the client that a keyspace is deleted and stop serving data access to that keyspace?

Reply:
The keyspace is valid throughout the session, which means we don't notify it.

Reply:
without notification, how to ensure the data can not be accessed when the keyspace is deleted?

Reply:
Notification would be a big change? For now clients can refresh their data? Would having a TTL for deletion help?

Even if a client is notified, what will stop them from using the deleted keyspace? We will have to follow this up with [auth](https://github.com/tikv/rfcs/pull/40) for keyspaces to enforce proper client behavior.

Reply:
updated. the client will refresh keyspace every 5min and return error if the keyspace is deleted.